README.md
Sparco
======

ChangeLog
2008-03-27  Ewout van den Berg  <ewout78@cs.ubc.ca>

	* generateProblem.m: fixed bug in handling additional parameters.

	* problems/prob901.m: fixed problem with thumbnail generation due
	to the change in problem data.

--------- Release 1.1.3 ------------------------------------------------

2008-03-25  Michael P. Friedlander <mpf@cs.ubc.ca>
	
	* tools/rwt: added pre-compiled MEX interfaces for Matlab
	- v.704 on Windows  (.dll)
	- v.7.5 (R2007b) on Linux (.mexglx)
	- v.7.5 (R2007b) on Windows (.mexw32)
	- v.7.5 (R2007b) on MacIntel (.mexmaci)
	(Thanks to Igor Carron for this suggestion.)  Also, changed the
	extensions of m-file placeholders to ".mhelp" so that older
	versions of Matlab have an easier time finding the .dll files.

        * examples/exSeismic: load seismic data from the 'data' subdir,
        and test for the CurveLab interface before running problem.

2008-03-24  Ewout van den Berg  <ewout78@cs.ubc.ca>

	* problems/data/prob502_angiogram.mat: store mask, pdf, and
	angiogram to avoid having to use 'twister' mode in rand

	* problems/prob502.m: load data from file for default
	problem. This avoids having to use the 'twister' mode in rand
	which was added only in Matlab 7.4. Thanks to Igor Carron for
	reporting this problem.

--------- Release 1.1.2 ------------------------------------------------

2008-03-21  Ewout van den Berg  <ewout78@cs.ubc.ca>

	* problems/data/prob901_seismic.mat: replaced normalized data by
	original data.

	* opCrop.m: Added  crop operator to `extract' rectangular regions
	from other operators.

	* opSum.m: Added sum operator to form the weighted sum of a set
	of operators.

	* opBlockDiag.m, opDictionary.m, opStack.m: Removed bug with
	respect to complex weights.

	* opToepGauss.m, opToepSign.m: Added Toeplitz-based measurement
	operators with Gaussian and sign entries.

	* dottest.m: Report ratio between y'*Ax and x'*A'y in case the dot
	test fails. This helps find scaling related bugs in operators.
	Thanks to Gilles Hennenfent for suggesting this feature.

	* opisreal.m: Added function that returns 1 if results of
	multiplication by the given operator are real, and 0
	otherwise. Also added an equivalent isreal member function to
	classOp.

2008-03-02  Ewout van den Berg  <ewout78@cs.ubc.ca>

	* opsize.m: Added opsize to query the size of an operator. This
	function works the same as the size function on matrices.

	* opStack.m: Added this operator to vertically stack a set of
	operators, similar to opDictionary which does this horizontally.
	
2008-02-15  Ewout van den Berg  <ewout78@cs.ubc.ca>

	* generateProblem.m: Added 'version' option to query Sparco
	version. Example: version = generateProblem('version');
	
2008-02-14  Ewout van den Berg  <ewout78@cs.ubc.ca>

	* operators/opHadamard.m: Added multiplication by the Hadamard
	matrix.

	* operators/opGaussian.m: Also added checks in row
	orthogonalization mode to see if number of rows does not exceed
	the number of columns, in which case the normalization cannot
	work.

	* problems/prob007.m, problems/prob008.m: Use QR instead of ORTH
	to generate orthogonal rows in A; speeds up problem generation.

2008-01-25  Ewout van den Berg  <ewout78@cs.ubc.ca>

	* problems/prob501.m: Bug fixed in generating the observation b.
	Thanks to Elaine Hale for reporting this bug.  

	* generateProblem.m: Added `lookup' option to get problem index
	based on name. Example: index = generateProblem(name/index,'lookup');

2008-01-23  Ewout van den Berg  <ewout78@cs.ubc.ca>

	* generateProblem.m: Added problem look up by name. Each data.info
	now also has a sparcoID field by default.
	Example: name = generateProblem(name/index,'getname');

	* parseDefaultOpts.m: Added parsing for getname flag.

	* problems/*: Added handling of the getname query.


2008-01-22  Ewout van den Berg  <ewout78@cs.ubc.ca>

	* operators/opSplitComplex.m: Added new non-linear operator for
	splitting complex numbers into its real and imaginary parts and
	vice versa.
	
2008-01-15  Michael P. Friedlander  <mpf@cs.ubc.ca>

	* problems/*: Added short names to all problems as info.name.

2008-01-12  Michael P. Friedlander  <mpf@cs.ubc.ca>

	* sparcoSetup: Don't compile RWT if the MEX interfaces already
	exist.

2007-11-27  Ewout van den Berg  <ewout78@cs.ubc.ca>

	* problems/prob001.m: Set the normalize flag in the Heaviside
	matrix to ensure the solution is indeed sparse. Without it there
	is too. much energy in the columns of the Heaviside matrix which
	gives a dense optimal solution.

	* problems/prob001.m: Generate the signal locally and compute the
	exact sparse solution.

--------- Release 1.1 --------------------------------------------------

2007-11-30  Ewout van den Berg  <ewout78@cs.ubc.ca>

	* operators/classOp/mtimes.m, operators/classOp/size.m: Support
	matrix and classOp operators in addition to functions. This
	feature can be used to unify matrix vector multiplication as well
	as to associated counters to matrices and existing classOp
	instances.


--------- Release 1.0 --------------------------------------------------


sparcoSetup.m
function sparcoSetup(varargin)
%SPARCOSETUP Setup the SPARCO toolbox.
%
%   SPARCOSETUP will install the SPARCO toolbox.

%   Copyright 2008, Ewout van den Berg and Michael P. Friedlander
%   http://www.cs.ubc.ca/labs/scl/sparco
%   $Id: sparcoSetup.m 1027 2008-06-24 23:42:28Z ewout78 $

pathlist = {};

% Get root location of sparcoSetup
root = fileparts(which(mfilename));

% ----------------------------------------------------------------------
% Add Sparco subdirs to path.
% ----------------------------------------------------------------------
addtopath(root,'');
addtopath(root,'examples');
%addtopath(root,'operators');
addtopath(root,'problems');
addtopath(root,'tools');
addtopath(root,['tools' filesep 'nufft']);
addtopath(root,['tools' filesep 'rwt']);

% Make sure that the build subdir exists
buildpath = {['build'],                          ...
             ['build' filesep 'figures'],        ...
             ['build' filesep 'html'],           ...
             ['build' filesep 'latex'],          ...
             ['build' filesep 'thumbs']};
for i=1:length(buildpath)
    p = buildpath{i};
    if ~exist([root filesep p],'dir')
       [success,message,messageid] = mkdir(root, p);
       if ~success
          error('Could not create directory %s%s\n',root,p);
       end
    end
end

% ----------------------------------------------------------------------
% Parse parameters.
% ----------------------------------------------------------------------
[opts,varg]= parseOptions(varargin,{'norwt'},{});

% % ----------------------------------------------------------------------
% % Check for external dependencies.
% % ----------------------------------------------------------------------
% 
% % Commented May 25th
% 
% % CurveLab: Req'd for 51-52
% if exist('curvelab.pdf','file')
%    crvroot = fileparts(which('curvelab.pdf'));
%    addtopath(crvroot,'fdct_usfft_matlab');
%    addtopath(crvroot,'fdct_wrapping_matlab');
%    addtopath(crvroot,'fdct_wrapping_cpp/mex');
%    addtopath(crvroot,'fdct3d/mex');
% else
%    fprintf(['\nWarning: CurveLab is not in the path. Problems 50-51 ' ...
%             'will not work.\n\n']);
% end
% 
% [T,result] = evalc('savepath;');
% rehash path;
% if result == 1
%    fprintf(['\n' ...
%             '\n Warning: SPARCO successfully added to your path,' ...
%             '\n          but couldn''t make these changes permanent.' ...
%             '\n          To permanently add SPARCO to your path,'...
%             '\n          copy and paste the following lines into your'...
%             '\n          startup.m file (e.g., ~/matlab/startup.m).'...
%             '\n\n']);
%    for i=1:length(pathlist)
%      fprintf(' addpath(''%s'');\n',pathlist{i});
%    end
%    fprintf('\n');
% end
% 
% % --'+sparco/',--------------------------------------------------------------------
% % Compile the Rice Wavelet Toolbox
% % ----------------------------------------------------------------------
% 
% if ~opts.norwt
%    cd([root filesep 'tools' filesep 'rwt'])
%    fprintf('Compiling the Rice Wavelet Toolbox MEX interfaces...');
%    try
%       if exist('mdwt'  ,'file')~=3, mex mdwt.c   mdwt_r.c;   end
%       if exist('midwt' ,'file')~=3, mex midwt.c  midwt_r.c;  end
%       if exist('mrdwt' ,'file')~=3, mex mrdwt.c  mrdwt_r.c;  end
%       if exist('mirdwNew Foldert','file')~=3, mex mirdwt.c mirdwt_r.c; end
%       fprintf('Success!\n');
%   catch
%       warning('Could not compile Rice Wavelet Toolbox MEX interfaces.');
%   end
%   cd(root)
% end

% ----------------------------------------------------------------------
% Check if Spot toolbox is installed and install it if not 
% ----------------------------------------------------------------------

isspot = strfind(ls,'spotbox');

if(~isempty(isspot))
    fprintf('Spot toolbox found\n')
else
    fprintf('Spot toolbox not found\n');
    fprintf('   Downloading Spot...\n');
    spotwebsite = 'http://www.cs.ubc.ca/labs/scl/spot/';
    website = urlread(spotwebsite);
    dl = strfind(website, 'Download');
    website = website(dl(1):size(website,2));
    spotURL = strfind(website, '<a href="');
    website = website((spotURL(1)+9):size(website,2));
    endofURL = strfind(website,'"');
    website = website(1:(endofURL-1));

    if(isempty(strfind(website,'www'))&&isempty(strfind(website,'http')))
        website = strcat(spotwebsite,website);
    end
    
    [path,status] = urlwrite(website,'spot.zip');

    if(status==0)
        warning('Could not download Spot toolbox from website\n');
    else
        fprintf('   Unzipping Spot toolbox...\n')
        unzip(path,'')
    end

    delete(path)

    list = dir;

    for i=1:length(list)
        if(~isempty(strfind(list(i).name,'spotbox')))
            if(list(i).isdir)
                spotDirName = list(i).name;
                newspotDirname = 'spotbox';
                movefile(spotDirName,newspotDirname);
                break
            end
        end
    end
end

addtopath(root,'spotbox');
% ----------------------------------------------------------------------
% NESTED FUNCTIONS
% ----------------------------------------------------------------------

function addtopath(root,dirname)
if ~exist([root filesep dirname],'dir')
   error('Required directory "%s%s%s" is missing!.\n',root,filesep,dirname);
else
   newdir = [root filesep dirname];
   fprintf('Adding to path: %s\n',newdir);
   addpath(newdir);
   pathlist{end+1} = newdir;
end
end % function adddirtopath

end % function sparcoSetup


% ----------------------------------------------------------------------
% PRIVATE FUNCTIONS
% ----------------------------------------------------------------------

function available = checkFunction(funname)

available = false;
w = which(funname);
if ~isempty(w)
   available = true;
   [root, name, ext, versn] = fileparts(w);
   if ~strcmp(name,funname)
      available = false;
      fprintf('Warning: Found case-insensitive match for %s.\n', funname);
   end
end

end % function checkFunction


